Fix: Stable implementation for ReduceLogSumExp (ONNX frontend)

sshekhar563 opened the pull request:
### Summary #32839 
This patch fixes numerical instability in the ONNX frontend’s `ReduceLogSumExp` operator.  
The previous implementation used a naïve `log(sum(exp(x)))` formulation, which overflowed to `Inf` for
values near `log(MAX_FLOAT32)` (≈ 88.72). This caused incorrect results in the OpenVINO EP of ONNX Runtime,
especially for float16 and float32 models.

### Fix
Implemented a numerically stable LogSumExp computation:

    k = ReduceMax(x)
    lse = k + log( ReduceSum( exp(x - k) ) )

This matches the behavior of:
- ONNX Runtime CPU EP  
- PyTorch and NumPy stable LogSumExp  
- OpenVINO PyTorch frontend implementation (`log.cpp`)

The new implementation is applied to:
- Opset 1–12 (`opset_1::reduce_log_sum_exp`)
- Opset 13–17 (`opset_13::reduce_log_sum_exp`)
- Opset 18+ (`opset_18::reduce_log_sum_exp`)

### Motivation
Fixes incorrect `Inf` outputs for values ≥ 88.7 when using OpenVINOExecutionProvider.  
The issue was originally reported here: *[link to GitHub issue]*.

### Validation
Tested using provided reproduction script from the issue:
- Matches ONNX Runtime CPU EP for all tested values
- No overflow observed for large positive inputs
- Behavior consistent across float16/float32 models

### Notes
This fix aligns ONNX frontend behavior with the already-correct PyTorch frontend implementation.

sshekhar563 committed:
Fix: stable implementation for ReduceLogSumExp (ONNX frontend)

cknd commented on src/frontends/onnx/frontend/src/op/reduce.cpp:
```diff
@@ -152,16 +154,29 @@ std::shared_ptr<ov::Node> onnx_reduce_sum_square(const ov::frontend::onnx::Node&
-    const ov::Output<ov::Node> sum_node =
-        make_ov_reduction_op<v1::ReduceSum>(node, node.get_ov_inputs().at(0), supported_types_v2);
-    return {std::make_shared<v0::Log>(sum_node)};
-}
```
is `reduce_log_sum` deleted on purpose?

sshekhar563 replied:
Thanks for the review!

The removal of reduce_log_sum was not intentional the change in this PR is meant only to introduce a stable implementation for ReduceLogSumExp.
I'll restore the original reduce_log_sum translator and its registration in the next commit.

Thanks for pointing it out!

cknd reviewed (commented):
Hey, thanks! 

We also looked into this in the meantime (our working version is here https://github.com/openvinotoolkit/openvino/compare/master...micropsi-industries:openvino:fix_logsumexp )
We were wondering about one or two points, see comments below

sshekhar563 commented on src/frontends/onnx/frontend/src/op/reduce.cpp:
```diff
@@ -303,10 +329,28 @@ ov::OutputVector reduce_min(const ov::frontend::onnx::Node& node) {
     return {make_ov_reduction_op<v1::ReduceMin>(node, node.get_ov_inputs().at(0), supported_types_v3, false)};
 }
 
-ov::OutputVector reduce_log_sum(const ov::frontend::onnx::Node& node) {
```
Thanks for pointing this out again!
The deletion of reduce_log_sum was not intentional I accidentally removed it while restructuring the file during the ReduceLogSumExp update.
I'll restore the original reduce_log_sum implementation and its registration in the next commit.
Thanks for catching this!

mvafin reviewed (changes requested):
Please return `reduce_log_sum` function and make a single helper function to create `ReduceLogSumExp` to reduce code duplication. Could you also add test that will show the numerical stability of this implementation?

sshekhar563 committed:
Refactor ReduceLogSumExp: restore reduce_log_sum, add helper, add stability tests

CuriousPanCake commented:
.build_jenkins

sshekhar563 committed:
Merge branch 'master' into fix-reduce-logsumexp

sshekhar563 committed:
Fix clang-format issues and add numerical stability test

sshekhar563 committed:
Merge branch 'fix-reduce-logsumexp' of https://github.com/sshekhar563/openvino into fix-reduce-logsumexp

sshekhar563 committed:
Merge branch 'master' into fix-reduce-logsumexp

sshekhar563 committed:
Fix: Stable implementation for ReduceLogSumExp (ONNX frontend)

- Fixed shape mismatch issue when keepdims=false by making both ReduceMax
  and ReduceSum operations use consistent keepdims parameter
- Eliminated conditional squeeze operation that caused fragile shape handling
- Maintains numerically stable log-sum-exp algorithm: k + log(sum(exp(x - k)))
- Added comprehensive tests for numerical stability with large values
- Resolves overflow issues for values >= 89.0 that would fail with naive implementation
- Supports both keepdims=true and keepdims=false configurations correctly

Fixes numerical stability and shape consistency issues in ReduceLogSumExp operation.

sshekhar563 committed:
Merge branch 'fix-reduce-logsumexp' of https://github.com/sshekhar563/openvino into fix-reduce-logsumexp

sshekhar563 committed:
Merge branch 'master' into fix-reduce-logsumexp

cknd commented:
This seems to be waiting for maintainer reviews

sshekhar563 committed:
Rename helper function to onnx_reduce_log_sum_exp_stable for consistency

sshekhar563 committed:
Merge branch 'master' into fix-reduce-logsumexp

cknd commented:
Could someone approve the CI workflows in order to help check for any remaining test failures?

sshekhar563 commented:
Hi maintainers
Just a gentle follow-up the CI workflows for this PR are still awaiting maintainer approval.
Could someone please approve them so the remaining checks can run?
Thanks a lot for your time and help!

sshekhar563 committed:
Simplify numerical stability test to pure C++ without runtime dependencies

sshekhar563 committed:
Apply clang-format fixes

cknd commented:
cc @mvafin @bumbosiepsak @CuriousPanCake

cknd commented:
Dear maintainers,

Can you please make it a priority to move this forward? This has been waiting for CI runs and final approval for several weeks; and there is a outdated change request by @mvafin that afaict has long since been adressed.

I'm not affiliated with the PR author, I just reported the bug #32839

cknd commented:
.

mvafin committed:
Merge branch 'master' into fix-reduce-logsumexp

mvafin commented on test_bug_32839_reproduction.py:
Please remove this file

mvafin commented on test_reduce_log_sum_exp.py:
Why is this file needed? Please remove it

sshekhar563 committed:
Address review feedback: fix broadcasting, remove Python scripts, fix headers

- Fix BLOCKER: ReduceMax now always uses keepdims=true internally for
  correct broadcasting during subtraction (input - k), then squeezes k
  at the end if keepdims=false. Matches PyTorch frontend pattern.
- Remove test_bug_32839_reproduction.py (per mvafin review)
- Remove test_reduce_log_sum_exp.py (per mvafin review)
- Add missing #include <vector> and <algorithm> in C++ test
- Fix copyright year 2025 -> 2026 in C++ test

sshekhar563 commented:
Hi @mvafin, thanks for the review! I've addressed all the feedback in the latest commit:

**Your requests:**
- ✅ Removed `test_bug_32839_reproduction.py`
- ✅ Removed `test_reduce_log_sum_exp.py`

**Copilot feedback also addressed:**
- ✅ **Fixed broadcasting bug (BLOCKER):** `ReduceMax` now always uses `keepdims=true` internally so that `input - k` broadcasts correctly for any reduction axis. If the original `keepdims` was false, `k` is squeezed at the end to match the output shape. This follows the same proven pattern used in the PyTorch frontend (`src/frontends/pytorch/src/op/log.cpp`).
- ✅ **Added missing headers:** `<vector>` and `<algorithm>` added to the C++ test file.
- ✅ **Fixed copyright year** (2025 → 2026).

Could you please re-review and approve the CI workflows when you get a chance? Thanks!

mvafin commented on src/frontends/onnx/tests/onnx_reduce_log_sum_exp_numerical_stability.cpp:
This test is not what is expected to test this change. We need a test that takes ONNX model provide inputs that will produce incorrect results with old implementation and verify that results are correct

sshekhar563 committed:
Replace standalone test with ONNX FE functional tests (mvafin review)

- Remove onnx_reduce_log_sum_exp_numerical_stability.cpp and its
  CMakeLists entry (mvafin: 'This test is not what is expected')
- Add proper ONNX FE functional tests in onnx_import.in.cpp that:
  * Use existing reduce_log_sum_exp.onnx models
  * Feed large inputs (100.0) above FP32 overflow threshold (~88.7)
  * Verify results are finite and correct (not Inf as with old impl)
  * Test both keepdims=true (opset 1-12) and keepdims=false (opset 18)
  * The opset 18 test also validates the broadcasting fix

sshekhar563 committed:
Merge branch 'master' into fix-reduce-logsumexp

sshekhar563 commented:
@mvafin Done removed the standalone C++ test and added proper ONNX FE functional tests in `onnx_import.in.cpp`:

- `onnx_model_reduce_log_sum_exp_large_values` — loads `reduce_log_sum_exp.onnx` (opset 1-12, keepdims=true), feeds inputs of 100.0 (well above FP32 overflow threshold ~88.7), verifies output = 102.77259. The old naive implementation would produce Inf here.
- `onnx_model_reduce_log_sum_exp_18_large_values` — loads `reduce_log_sum_exp_18.onnx` (opset 18, keepdims=false), same large inputs, verifies output = 101.38629. This also validates the broadcasting fix for non-trailing axis reduction.

Also merged upstream master to resolve the out-of-date branch CI failures.

mvafin commented:
build_jenkins

mvafin commented:
@sshekhar563 Please fix code style

sshekhar563 committed:
Fix clang-format violation in test code

sshekhar563 commented:
@mvafin Fixed the code style in commit aa4cc53. Verified locally with clang-format all files pass clean.
Could you please take another look so we can proceed with full CI validation? Thanks!

mvafin commented:
build_jenkins

sshekhar563 committed:
Fix supported_ops.md: add missing opset 13 for ReduceLogSum

sshekhar563 committed:
Merge upstream master into fix-reduce-logsumexp

sshekhar563 commented:
@mvafin The clang-format CI was failing because `supported_ops.md` was out of sync with the code registrations. My earlier commits added an opset 13 handler for `ReduceLogSum` (alongside the `ReduceLogSumExp` fix), but I missed updating the docs table to reflect that.
I've updated `supported_ops.md` to change `ReduceLogSum` supported versions from `18, 1` → `18, 13, 1` to match the actual registrations in `reduce.cpp`.

Could you confirm this is correct? If the `ReduceLogSum` opset 13 registration itself shouldn't be part of this PR, I'm happy to revert both the registration and the docs change to keep the scope focused on `ReduceLogSumExp` only.

mvafin commented:
build_jenkins

cknd commented:
any way we can help?

mvafin committed:
Merge branch 'master' into fix-reduce-logsumexp

mvafin commented:
build_jenkins
